lib/bolt/transport/lxd/connection.rb

```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+require 'logging'
+require 'bolt/node/errors'
+
+module Bolt
+  module Transport
+    class LXD < Simple
+      class Connection
+        attr_reader :user, :target
+
+        def initialize(target, options)
+          @target = target
+          @user = ENV['USER'] || Etc.getlogin
+          @options = options
+          @lxd_remote = @target.config.dig('lxd', 'remote') || default_remote
+          @logger = Bolt::Logger.logger(target.safe_name)
+          @logger.trace("Initializing LXD connection to #{target.safe_name}")
+        end
+
+        def shell
+          Bolt::Shell::Bash.new(target, self)
+        end
+
+        def container_id
+          "#{@lxd_remote}:#{@target.uri}"
+        end
+
+        def connect
+          # TODO: check container is running
+          # TODO: get info about container, store on self
+          # @container_info = get_info
+          true
+        rescue StandardError => e
+          raise Bolt::Node::ConnectError.new(
+            "Failed to connect to #{container_id}: #{e.message}",
+            'CONNECT_ERROR'
+          )
+        end
+
+        def execute(command)
+          env_vars = []
+          if command.start_with?("PT_")
+            parts = Shellwords.split(command)
+            env_vars, command = parts.partition { |p| p.start_with?("PT_") }
+            command = Shellwords.shelljoin(command)
+          end
+
+          args = []
+          # See `lxc exec --help` for information on flags`
+          env_vars.each do |env_var|
+            args += %W[--env #{env_var}]
+          end
+
+          lxc_command = %w[lxc exec] + args + %W[#{container_id} --] + Shellwords.split(command)
+
+          @logger.trace { "Executing: #{lxc_command.join(' ')}" }
+          Open3.popen3(*lxc_command)
+        end
+
+        private def execute_local_command(command)
+          Open3.capture3('lxc', *command, { binmode: true })
+        end
+
+        def upload_file(source, destination)
+          @logger.trace { "Uploading #{source} to #{destination}" }
+          cmd = %W[file push --recursive #{source} #{container_id}:#{destination}]
+          _stdout, stderr, status = execute_local_command(cmd)
+          unless status.exitstatus.zero?
+            raise "Error writing to #{container_id}: #{stderr}"
+          end
+        rescue StandardError => e
+          raise Bolt::Node::FileError.new(e.message, 'WRITE_ERROR')
+        end
+
+        def download_file(source, destination, _download)
+          @logger.trace { "Downloading #{source} to #{destination}" }
+          FileUtils.mkdir_p(destination)
+          _stdout, stderr, status = execute_local_command(%W[file pull #{container_id}:#{source} destination])
```

Comment:
```suggestion
          _stdout, stderr, status = execute_local_command(%W[file pull #{container_id}#{source} destination])
```
Very similar to the previous comment, but this time for `file pull`

```
❯ lxc file pull --help
Description:
  Pull files from instances

Usage:
  lxc file pull [<remote>:]<instance>/<path> [[<remote>:]<instance>/<path>...] <target path> [flags]

Examples:
  lxc file pull foo/etc/hosts .
     To pull /etc/hosts from the instance and write it to the current directory.
```